### PR TITLE
Performance miniapps: don't enable -march=native on ARM

### PR DESCRIPTION
**Pull request by matthiasdiener**
See e.g. https://stackoverflow.com/questions/65966969/why-does-march-native-not-work-on-apple-m1

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2273](https://github.com/mfem/mfem/pull/2273) | @matthiasdiener | @tzanio | @tzanio + @camierjs | 05/27/21 | 05/28/21 | 05/30/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
Hi @matthiasdiener,

Thank you for the contribution!

If this is ready to be reviewed, can you please mark it with the `ready-for-review` label.

See also [CONTRIBUTING.md](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md).

Cheers,
Tzanio

**Comment by matthiasdiener** [comment]
> Hi @matthiasdiener,
> 
> Thank you for the contribution!
> 
> If this is ready to be reviewed, can you please mark it with the `ready-for-review` label.
> 
> See also [CONTRIBUTING.md](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md).
> 
> Cheers,
> Tzanio

Thanks @tzanio! This PR is ready for review, but I do not have permissions to add a label. Do I need to join the mfem organization first?

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by tzanio — APPROVED** [review]
I don't have a way to test it, but it looks good. Thanks!

**Comment by tzanio** [comment]
@samuelpmishLLNL do you think you can help us with this (`make test` in `miniapps/performance`)

**Review by camierjs — APPROVED** [review]

File: miniapps/performance/makefile [hunk 1]
```diff
@@ -98,7 +98,10 @@ MFEM_PERF_CXXFLAGS_gcc_ppc64 = -mcpu=native -mtune=native\
 MFEM_PERF_CXXFLAGS_xlc = -mcpu=native
 
 # - Clang extra options:
-MFEM_PERF_CXXFLAGS_clang += -march=native
+ifneq ($(MFEM_MACHINE),arm64)
+  # -march=native is unavailable for ARM
```
[the remark this answers is not in the document]
Comment (matthiasdiener) [reply miniapps/performance/makefile]: Good point, I clarified the comment.

**Comment by samuelpmishLLNL** [comment]
> @samuelpmishLLNL do you think you can help us with this? (`make test` in `miniapps/performance`)

I'm not too familiar with the mfem build process, here's what I tried:

```
code % git clone git@github.com:matthiasdiener/mfem.git && cd mfem
  ...
mfem % git checkout arm64_clang_flag
  ...
mfem % cd miniapps/performance && make test
  makefile:151: *** The MFEM library is not built.  Stop.
performance % cd ../../ && make config
  /Applications/Xcode.app/Contents/Developer/usr/bin/make -C config all
  Build destination: <source> [/Users/sam/code/mfem]
mfem % make -j
  ...
  a - fem/tmop/tmop_pa_tc3.o
  a - fem/tmop/tmop_pa_w2.o
  a - fem/tmop/tmop_pa_w2_c0.o
  a - fem/tmop/tmop_pa_w3.o
  a - fem/tmop/tmop_pa_w3_c0.o
  ranlib -no_warning_for_no_symbols libmfem.a
mfem % cd miniapps/performance && make test

  Detected host compiler: clang
  g++  -O3 -std=c++11 -pedantic -Wall -fcolor-diagnostics -fvectorize -fslp-vectorize -ffp-contract=fast -I../..  ex1.cpp -o ex1 -L../.. -lmfem 
      Performance miniapp [ ex1 ... ]: FAILED  (0.45s 8048kB)
  Options used:
     --mesh ../../data/fichera.mesh
     --refine 2
     --order 3
     --basis-type G
     --hpc-version
     --matrix-free
     --preconditioner none
     --no-static-condensation
     --no-visualization
  
  MFEM SIMD width: 1 doubles
  
  Using Gauss-Lobatto basis ...
  High-performance version using integration rule with 125 points ...
  Switching the mesh curvature to match the optimized value (order 3) ...
  Number of finite element unknowns: 13897
  Assembling the bilinear form ... done, 0.000974167s.
  Size of linear system: 13897
  Assembling the preconditioning matrix ... done, 1.66667e-07s.
     Iteration :   0  (B r, r) = 0.00558641
     Iteration :   1  (B r, r) = 0.0120229
     Iteration :   2  (B r, r) = 0.0127274
     Iteration :   3  (B r, r) = 0.018909
     Iteration :   4  (B r, r) = 0.0141179
     Iteration :   5  (B r, r) = 0.00948901
     Iteration :   6  (B r, r) = 0.00384171
     Iteration :   7  (B r, r) = 0.00537885
     Iteration :   8  (B r, r) = 0.00408867
     Iteration :   9  (B r, r) = 0.00296062
     Iteration :  10  (B r, r) = 0.00174241
     Iteration :  11  (B r, r) = 0.00113632
     Iteration :  12  (B r, r) = 0.000809106
     Iteration :  13  (B r, r) = 0.000858512
     Iteration :  14  (B r, r) = 0.00059977
     Iteration :  15  (B r, r) = 0.000450216
     Iteration :  16  (B r, r) = 0.000238495
     Iteration :  17  (B r, r) = 0.000122265
     Iteration :  18  (B r, r) = 7.51681e-05
     Iteration :  19  (B r, r) = 7.7508e-05
     Iteration :  20  (B r, r) = 3.60636e-05
     Iteration :  21  (B r, r) = 1.72552e-05
     Iteration :  22  (B r, r) = 9.85336e-06
     Iteration :  23  (B r, r) = 1.1665e-05
     Iteration :  24  (B r, r) = 8.85087e-06
     Iteration :  25  (B r, r) = 8.10593e-06
     Iteration :  26  (B r, r) = 5.18349e-06
     Iteration :  27  (B r, r) = 2.62407e-06
     Iteration :  28  (B r, r) = 2.01851e-06
     Iteration :  29  (B r, r) = 1.35531e-06
     Iteration :  30  (B r, r) = 8.90586e-07
     Iteration :  31  (B r, r) = 4.89596e-07
     Iteration :  32  (B r, r) = 2.51159e-07
     Iteration :  33  (B r, r) = 1.20681e-07
     Iteration :  34  (B r, r) = 9.74755e-08
     Iteration :  35  (B r, r) = 7.70346e-08
     Iteration :  36  (B r, r) = 4.35101e-08
     Iteration :  37  (B r, r) = 1.8659e-08
     Iteration :  38  (B r, r) = 1.17833e-08
     Iteration :  39  (B r, r) = 5.35768e-09
     Iteration :  40  (B r, r) = 3.15724e-09
     Iteration :  41  (B r, r) = 1.93656e-09
     Iteration :  42  (B r, r) = 7.97049e-10
     Iteration :  43  (B r, r) = 3.36088e-10
     Iteration :  44  (B r, r) = 2.80251e-10
     Iteration :  45  (B r, r) = 2.6404e-10
     Iteration :  46  (B r, r) = 1.2283e-10
     Iteration :  47  (B r, r) = 8.59806e-11
     Iteration :  48  (B r, r) = 4.87587e-11
     Iteration :  49  (B r, r) = 2.57866e-11
     Iteration :  50  (B r, r) = 1.76063e-11
     Iteration :  51  (B r, r) = 1.4736e-11
     Iteration :  52  (B r, r) = 7.39069e-12
     Iteration :  53  (B r, r) = 4.57616e-12
     Iteration :  54  (B r, r) = 2.44942e-12
     Iteration :  55  (B r, r) = 1.88052e-12
     Iteration :  56  (B r, r) = 1.41005e-12
     Iteration :  57  (B r, r) = 8.05961e-13
     Iteration :  58  (B r, r) = 3.6259e-13
     Iteration :  59  (B r, r) = 2.01262e-13
     Iteration :  60  (B r, r) = 1.62509e-13
     Iteration :  61  (B r, r) = 7.15582e-14
     Iteration :  62  (B r, r) = 3.51032e-14
     Iteration :  63  (B r, r) = 1.86634e-14
     Iteration :  64  (B r, r) = 8.82869e-15
     Iteration :  65  (B r, r) = 3.72374e-15
  Average reduction factor = 0.806004
  make: *** [ex1-test-seq] Error 252
  make: Target `test' not remade because of errors.
```

**Comment by matthiasdiener** [comment]
> > @samuelpmishLLNL do you think you can help us with this? (`make test` in `miniapps/performance`)
> 
> I'm not too familiar with the mfem build process, here's what I tried:
> 
> ```
> code % git clone git@github.com:matthiasdiener/mfem.git && cd mfem
>   ...
> mfem % git checkout arm64_clang_flag
>   ...
> mfem % cd miniapps/performance && make test
>   makefile:151: *** The MFEM library is not built.  Stop.
> performance % cd ../../ && make config
>   /Applications/Xcode.app/Contents/Developer/usr/bin/make -C config all
>   Build destination: <source> [/Users/sam/code/mfem]
> mfem % make -j
>   ...
>   a - fem/tmop/tmop_pa_tc3.o
>   a - fem/tmop/tmop_pa_w2.o
>   a - fem/tmop/tmop_pa_w2_c0.o
>   a - fem/tmop/tmop_pa_w3.o
>   a - fem/tmop/tmop_pa_w3_c0.o
>   ranlib -no_warning_for_no_symbols libmfem.a
> mfem % cd miniapps/performance && make test
> 
>   Detected host compiler: clang
>   g++  -O3 -std=c++11 -pedantic -Wall -fcolor-diagnostics -fvectorize -fslp-vectorize -ffp-contract=fast -I../..  ex1.cpp -o ex1 -L../.. -lmfem 
>       Performance miniapp [ ex1 ... ]: FAILED  (0.45s 8048kB)
>   Options used:
>      --mesh ../../data/fichera.mesh
>      --refine 2
>      --order 3
>      --basis-type G
>      --hpc-version
>      --matrix-free
>      --preconditioner none
>      --no-static-condensation
>      --no-visualization
>   
>   MFEM SIMD width: 1 doubles
>   
>   Using Gauss-Lobatto basis ...
>   High-performance version using integration rule with 125 points ...
>   Switching the mesh curvature to match the optimized value (order 3) ...
>   Number of finite element unknowns: 13897
>   Assembling the bilinear form ... done, 0.000974167s.
>   Size of linear system: 13897
>   Assembling the preconditioning matrix ... done, 1.66667e-07s.
>      Iteration :   0  (B r, r) = 0.00558641
>      Iteration :   1  (B r, r) = 0.0120229
>      Iteration :   2  (B r, r) = 0.0127274
>      Iteration :   3  (B r, r) = 0.018909
>      Iteration :   4  (B r, r) = 0.0141179
>      Iteration :   5  (B r, r) = 0.00948901
>      Iteration :   6  (B r, r) = 0.00384171
>      Iteration :   7  (B r, r) = 0.00537885
>      Iteration :   8  (B r, r) = 0.00408867
>      Iteration :   9  (B r, r) = 0.00296062
>      Iteration :  10  (B r, r) = 0.00174241
>      Iteration :  11  (B r, r) = 0.00113632
>      Iteration :  12  (B r, r) = 0.000809106
>      Iteration :  13  (B r, r) = 0.000858512
>      Iteration :  14  (B r, r) = 0.00059977
>      Iteration :  15  (B r, r) = 0.000450216
>      Iteration :  16  (B r, r) = 0.000238495
>      Iteration :  17  (B r, r) = 0.000122265
>      Iteration :  18  (B r, r) = 7.51681e-05
>      Iteration :  19  (B r, r) = 7.7508e-05
>      Iteration :  20  (B r, r) = 3.60636e-05
>      Iteration :  21  (B r, r) = 1.72552e-05
>      Iteration :  22  (B r, r) = 9.85336e-06
>      Iteration :  23  (B r, r) = 1.1665e-05
>      Iteration :  24  (B r, r) = 8.85087e-06
>      Iteration :  25  (B r, r) = 8.10593e-06
>      Iteration :  26  (B r, r) = 5.18349e-06
>      Iteration :  27  (B r, r) = 2.62407e-06
>      Iteration :  28  (B r, r) = 2.01851e-06
>      Iteration :  29  (B r, r) = 1.35531e-06
>      Iteration :  30  (B r, r) = 8.90586e-07
>      Iteration :  31  (B r, r) = 4.89596e-07
>      Iteration :  32  (B r, r) = 2.51159e-07
>      Iteration :  33  (B r, r) = 1.20681e-07
>      Iteration :  34  (B r, r) = 9.74755e-08
>      Iteration :  35  (B r, r) = 7.70346e-08
>      Iteration :  36  (B r, r) = 4.35101e-08
>      Iteration :  37  (B r, r) = 1.8659e-08
>      Iteration :  38  (B r, r) = 1.17833e-08
>      Iteration :  39  (B r, r) = 5.35768e-09
>      Iteration :  40  (B r, r) = 3.15724e-09
>      Iteration :  41  (B r, r) = 1.93656e-09
>      Iteration :  42  (B r, r) = 7.97049e-10
>      Iteration :  43  (B r, r) = 3.36088e-10
>      Iteration :  44  (B r, r) = 2.80251e-10
>      Iteration :  45  (B r, r) = 2.6404e-10
>      Iteration :  46  (B r, r) = 1.2283e-10
>      Iteration :  47  (B r, r) = 8.59806e-11
>      Iteration :  48  (B r, r) = 4.87587e-11
>      Iteration :  49  (B r, r) = 2.57866e-11
>      Iteration :  50  (B r, r) = 1.76063e-11
>      Iteration :  51  (B r, r) = 1.4736e-11
>      Iteration :  52  (B r, r) = 7.39069e-12
>      Iteration :  53  (B r, r) = 4.57616e-12
>      Iteration :  54  (B r, r) = 2.44942e-12
>      Iteration :  55  (B r, r) = 1.88052e-12
>      Iteration :  56  (B r, r) = 1.41005e-12
>      Iteration :  57  (B r, r) = 8.05961e-13
>      Iteration :  58  (B r, r) = 3.6259e-13
>      Iteration :  59  (B r, r) = 2.01262e-13
>      Iteration :  60  (B r, r) = 1.62509e-13
>      Iteration :  61  (B r, r) = 7.15582e-14
>      Iteration :  62  (B r, r) = 3.51032e-14
>      Iteration :  63  (B r, r) = 1.86634e-14
>      Iteration :  64  (B r, r) = 8.82869e-15
>      Iteration :  65  (B r, r) = 3.72374e-15
>   Average reduction factor = 0.806004
>   make: *** [ex1-test-seq] Error 252
>   make: Target `test' not remade because of errors.
> ```

Yeah, the test will still fail on an M1, but at least it will build.

**Comment by tzanio** [comment]
One sec, guys we have a fix...

**Comment by tzanio** [comment]
@matthiasdiener, can you pull `master` and merge it in this branch... I was not able to do it

**Comment by matthiasdiener** [comment]
> @matthiasdiener, can you pull `master` and merge it in this branch... I was not able to do it

I *think* I it.

**Comment by tzanio** [comment]
Ok, can you pull and try `make test` in `miniapps/performance` again? (no need to rebuild)

**Comment by samuelpmishLLNL** [comment]
update:

```
performance % make test                           
    Performance miniapp [ ex1 ... ]: OK  (0.09s 7904kB)
```

**Comment by tzanio** [comment]
Thank you! The fix was for Big Sur, not M1 specifically, see #2276

**Pull request merged**